judge/models/submission.py

```diff
@@ -121,18 +121,47 @@ def long_status(self):
     def is_locked(self):
         return self.locked_after is not None and self.locked_after < timezone.now()
 
-    def judge(self, *args, rejudge=False, force_judge=False, rejudge_user=None, **kwargs):
+    def judge(self, *, rejudge=False, force_judge=False, rejudge_user=None, **kwargs):
         if force_judge or not self.is_locked:
             if rejudge:
                 with revisions.create_revision(manage_manually=True):
                     if rejudge_user:
                         revisions.set_user(rejudge_user)
                     revisions.set_comment('Rejudged')
                     revisions.add_to_revision(self)
-            judge_submission(self, *args, rejudge=rejudge, **kwargs)
+            judge_submission(self, rejudge=rejudge, **kwargs)
 
     judge.alters_data = True
 
+    @classmethod
+    def batch_judge(cls, submissions, *, rejudge=False, force_judge=False, rejudge_user=None, batch_size=100, **kwargs):
+        # Don't trust the caller to follow related objects that we need.
+        submissions = submissions.select_related('problem', 'language', 'source')
+
+        if not force_judge:
+            submissions = submissions.exclude(locked_after__lt=timezone.now())
+
+        # Force evaluate the queryset here so we have a fixed set of submissions that we are working with
+        # rather than a lazy queryset that could change.
+        submissions = list(submissions)
```

Comment:
Thought about using `.iterator()` instead of forcefully evaluating and caching the entire queryset so we don't use too much memory, but not sure if it's worth it given that MySQL doesn't support streaming results and so the entire result set would be loaded anyways ([Source](https://docs.djangoproject.com/en/2.2/ref/models/querysets/#without-server-side-cursors)). The benefit is that we wouldn't be storing expensive Python objects, but it would complicate this code more.

Reply:
No, let's use `.iterator()`. Let me create a quick chunk helper function and you can use it then.

Reply:
Pulled in #1903 temporarily for completeness.